crates/revm/src/processor.rs

```diff
@@ -0,0 +1,511 @@
+            // must be no greater than the block’s gasLimit.
+            let block_available_gas = block.header.gas_limit - cumulative_gas_used;
+            if transaction.gas_limit() > block_available_gas {
+                return Err(BlockValidationError::TransactionGasLimitMoreThanAvailableBlockGas {
+                    transaction_gas_limit: transaction.gas_limit(),
+                    block_available_gas,
+                }
+                .into())
+            }
+            // Execute transaction.
+            let ResultAndState { result, state } = self.transact(transaction, sender)?;
+            trace!(
+                target: "evm",
+                ?transaction, ?result, ?state,
+                "Executed transaction"
+            );
+            self.stats.execution_duration += time.elapsed();
+            let time = Instant::now();
+
+            self.db().commit(state);
+
+            self.stats.apply_state_duration += time.elapsed();
+
+            // append gas used
+            cumulative_gas_used += result.gas_used();
+
+            // Push transaction changeset and calculate header bloom filter for receipt.
+            receipts.push(Receipt {
+                tx_type: transaction.tx_type(),
+                // Success flag was added in `EIP-658: Embedding transaction status code in
+                // receipts`.
+                success: result.is_success(),
+                cumulative_gas_used,
+                // convert to reth log
+                logs: result.into_logs().into_iter().map(into_reth_log).collect(),
+            });
+        }
+
+        Ok((receipts, cumulative_gas_used))
+    }
+
+    /// Execute the block, verify gas usage and apply post-block state changes.
+    fn execute_inner(
+        &mut self,
+        block: &Block,
+        total_difficulty: U256,
+        senders: Option<Vec<Address>>,
+    ) -> Result<Vec<Receipt>, BlockExecutionError> {
+        let (receipts, cumulative_gas_used) =
+            self.execute_transactions(block, total_difficulty, senders)?;
+
+        // Check if gas used matches the value set in header.
+        if block.gas_used != cumulative_gas_used {
+            return Err(BlockValidationError::BlockGasUsed {
+                got: cumulative_gas_used,
+                expected: block.gas_used,
+                gas_spent_by_tx: self
+                    .receipts
+                    .last()
+                    .map(|block_r| {
+                        block_r
+                            .iter()
+                            .enumerate()
+                            .map(|(id, tx_r)| {
+                                (
+                                    id as u64,
+                                    tx_r.as_ref()
+                                        .expect("receipts have not been pruned")
+                                        .cumulative_gas_used,
+                                )
```

Comment:
would consider making a method on receipts

Reply:
Can be addressed by `Receipts` type: https://github.com/paradigmxyz/reth/issues/4621